### PR TITLE
[KV Offload] Enable KV offloading for attention data parallelism

### PR DESCRIPTION
**Pull request by amitkumar307d**
# Description

This PR enables KV Cache Offloading support when Attention Data Parallelism (`enable_dp_attention: true`) is enabled.

**Context & Problem:**
Previously, when KV offloading was used alongside Attention DP, the system relied on custom Pallas ops (`stack_kv_cache_cross_layers` and `update_kv_caches_one`) that did not support the `attn_dp` sharding layout. This led to errors during the save/load chunk operations. 

**Error:**
```bash
(EngineCore pid=533) WARNING 06-04 15:37:02 [tpu_offload_connector.py:1433]     - Failed to pre-compile for 1 blocks: shard_map applied to the function '_start_chunked_copy_kernel' was given argument arrays with axis sizes that are not evenly divisible by the corresponding mesh axis sizes:
(EngineCore pid=533) WARNING 06-04 15:37:02 [tpu_offload_connector.py:1433] 
(EngineCore pid=533) WARNING 06-04 15:37:02 [tpu_offload_connector.py:1433] The mesh given has shape (1, 8, 1, 1, 1, 1) with corresponding axis names ('data', 'attn_dp', 'attn_dp_expert', 'expert', 'model', 'dcp').
(EngineCore pid=533) WARNING 06-04 15:37:02 [tpu_offload_connector.py:1433] 
(EngineCore pid=533) WARNING 06-04 15:37:02 [tpu_offload_connector.py:1433] the passed args[4] of shape float8_e4m3fn[1,256,4,640], where args[4] is bound to _start_chunked_copy_kernel's parameter 'src_array: jax.Array', corresponds to in_specs[4] of value P(('data', 'attn_dp', 'attn_dp_expert'), 'dcp'), which maps array axis 0 (of size 1) to mesh axes ('data', 'attn_dp', 'attn_dp_expert') (of total size 8), but 8 does not evenly divide 1
(EngineCore pid=533) WARNING 06-04 15:37:02 [tpu_offload_connector.py:1433] 
(EngineCore pid=533) WARNING 06-04 15:37:02 [tpu_offload_connector.py:1433] Array arguments' axis sizes must be evenly divisible by the mesh axis or axes indicated by the corresponding elements of the argument's in_specs entry. Consider checking that in_specs are correct, and if so consider changing the mesh axis sizes or else padding the input and adapting '_start_chunked_copy_kernel' appropriately.
```

# Solution
- **Pure JAX Fallbacks:** Added pure JAX implementations (`pure_jax_stack_kv_cache_cross_layers` and `pure_jax_update_kv_caches_one`) in `offload/utils.py` using `jax.lax.scatter`.
- **Dynamic Dispatch:** Updated `TPUOffloadConnectorWorker` to conditionally route to these pure JAX implementations via `_do_stack` and `_do_update` when DP attention is enabled.
- **Metrics Fixes:** 
  - Fixed a bug in `KVOffloadConnectorStats.aggregate()` where it returned without accumulating the nested chunk dictionaries, permanently dropping metrics.
  - Fixed `TPUOffloadConnector.get_kv_connector_stats()` to correctly delegate to `self.connector_worker` when executing in the worker role, restoring metrics logging for DeepSeek.
- **DeepSeek V4 Sync:** Reconciled integration conflicts with upstream DeepSeek V4 changes in `kv_cache_manager.py` and `deepseek_v4_attention.py`.

# Tests
- Verified that KV cache offloading executes successfully without sharding crashes when `--additional_config='{"sharding": {"sharding_strategy": {"enable_dp_attention": true}}}'` is passed.
- Verified that `KV Transfer metrics` are now correctly aggregated and printed in the logs for DeepSeek inference runs.

**Tested Configuration Example:**
```bash
--additional_config='{"sharding": {"sharding_strategy": {"enable_dp_attention": true}}}' \
--kv-transfer-config='{"kv_connector": "TPUOffloadConnector", "kv_connector_module_path": "tpu_inference.offload.tpu_offload_connector", "kv_role": "kv_both", "kv_connector_extra_config": {"cpu_bytes_to_use": 107374182400, "lazy_offload": true}}'
```

**Deepseek V3.1 671b - Server command:**
```bash
python3 -m vllm.entrypoints.openai.api_server \
  --host=0.0.0.0 \
  --port=8000 \
  --download-dir=/data \
  --kv-cache-dtype=fp8 \
  --max-model-len=20480 \
  --gpu-memory-utilization=0.98 \
  --tensor-parallel-size=8 \
  --max-num-seqs=128 \
  --max-num-batched-tokens=512 \
  --enable-prefix-caching \
  --enable-expert-parallel \
  --async-scheduling \
  --model=deepseek-ai/DeepSeek-V3.1 \
  --served-model-name=deepseek-ai/DeepSeek-V3.1 \
  --load-format=runai_streamer \
  --additional_config='{"sharding": {"sharding_strategy": {"enable_dp_attention": true}}}' \
  --kv-transfer-config='{"kv_connector": "TPUOffloadConnector", "kv_connector_module_path": "tpu_inference.offload.tpu_offload_connector","kv_role": "kv_both", "kv_connector_extra_config": {"cpu_bytes_to_use": 107374182400, "lazy_offload": false}}' || true
```

**Deepseek V3.1 671b - Benchmark command:**
```bash
vllm bench serve \
  --backend=openai \
  --model=deepseek-ai/DeepSeek-V3.1 \
  --dataset-name=prefix_repetition \
  --host=localhost \
  --port=8000 \
  --seed=123 \
  --num-prompts=32 \
  --max-concurrency=32 \
  --prefix-repetition-prefix-len=19424 \
  --prefix-repetition-suffix-len=32 \
  --prefix-repetition-output-len=1024 \
  --prefix-repetition-num-prefixes=4 \
  --percentile-metrics='ttft,tpot,itl,e2el' \
  --ignore-eos
```

# Results

| Metric | With KV Offloading | Without KV Offloading |
| :--- | :--- | :--- |
| **Output Token Throughput (tok/s)** | 506.75 (+67.4%) | 302.68 |
| **Total Token Throughput (tok/s)** | 10134.95 (+67.4%) | 6053.55 |
| **Mean TTFT (ms)** | 2958.56 (-86.8%) | 22472.25 |
| **Mean TPOT (ms)** | 39.41 (+7.7%) | 36.59 |
| **Mean ITL (ms)** | 39.64 (+8.3%) | 36.59 |
| **Mean E2EL (ms)** | 43269.93 (-27.8%) | 59907.64 |

# Checklist
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

**Comment by amitkumar307d** [comment]
@vllm-project/tpu-inference-maintainers 

Could a maintainer please add the "ready" label to this PR? I don't currently have the triage/push permissions needed to add labels myself. 

This PR is ready for review and addresses KV Cache Offloading support with Attention Data Parallelism.

Thank you!

**Comment by amitkumar307d** [comment]
@juncgu-google could you also please review this PR?

**Comment by juncgu-google** [comment]
> @juncgu-google could you also please review this PR?

Thanks, @amitkumar307d.
My two cents (high-level suggestions) to this PR:
1. have you tested / micro-bencmarked the performance of `pure_jax_update_kv_caches_one` ? My experience is jax-based kv update approach may import extra-large latency numbers. Furthermore, do you have e2d performance numbers of baseline (without KV offloading) to compare against the numbers shown above?
2. It will be good to add unit tests for the major changes (e.g., `pure_jax_update_kv_caches_one` and `pure_jax_stack_kv_cache_cross_layers` ) in this PR.

**Comment by amitkumar307d** [comment]
@juncgu-google Thanks for reviewing. 

- That's a valid callout. Because pure_jax uses @jax.jit on a list of blocks, dynamic fluctuations in list length from the scheduler will trigger JIT recompilations, causing those latency spikes. However, we used **pure JAX** to sepcifically satisfy the mesh constraints for **attn_dp parallelism**, and our end-to-end benchmarks show acceptable performance. If future micro-benchmarks show severe regressions, we can optimise this by padding the block lists to a static shape to eliminate recompilations.
- Updated the PR description with the end-to-end benchmark markdown table comparing the performance both with and without KV offloading on the DeepSeek benchmark, so the baseline numbers are there for review! 
- Unit tests added for the newly added methods in utils.

File: tpu_inference/offload/tpu_offload_connector.py [hunk 1]
```diff
@@ -1386,6 +1440,18 @@ def _decompose_into_buckets(self, block_ids: list[int]) -> list[list[int]]:
                 raise ValueError("Could not decompose with given buckets.")
         return decomposed_blocks
 
+    def _do_stack(self, kv_caches, block_ids_arr, num_blocks):
+        if self.mesh.shape.get('attn_dp', 1) > 1:
```
Comment (juncgu-google) [inline tpu_inference/offload/tpu_offload_connector.py:1451]: could we abstract this line into a function (e.g., `is_attn_dp`)? or there might be such a function exist, if not, please put the function at utils.py

**Review by juncgu-google — APPROVED** [review]
LGTM.

File: tpu_inference/offload/tpu_offload_connector.py (hunk 1 above)
Comment (amitkumar307d) [reply tpu_inference/offload/tpu_offload_connector.py:1451]: We can but I kept the actual _do_stack and _do_update wrappers inside the Connector class so that the Connector remains responsible for routing based on its own state (self.mesh), while utils.py stays purely as a stateless math toolbox.

**Review by kyuyeunk — APPROVED** [review]

**Review by sixiang-google — APPROVED** [review]
